### ISSUE TITLE
nix-env -r deprecation

### ISSUE THREAD

**Issue by bryaan** [NixOS/nix#1866]
`nix-env -iA nixpkgs.userEnv`

Error: some pacakge is already installed and need to be removed first.  Ok, makes sense:

`nix-env -iAr nixpkgs.userEnv`

(Nix newb:) So this should remove all the packages I don't need and install my stuff.

What!? where is nix-env?? The -r just removed it all.  Back to nix homepage to reinstall and thanks higher power for dotfiles repo.

So hopefully this illustrates why -r is absolutely terrible and need to be deprecated ASAP.

This also violates principles of least surprise.

If you want to run such a destructive operation it most certainly shouldn't be a *single keystroke* to trigger.  It should be the explicit:  `nix-env -e '.*' `

**Comment by shlevy** [comment]
We are working on a better declarative user env story, one that hopefully will avoid this issue, but we shouldn't deprecate what currently exists until there's a better replacement. I agree what happened here was unfortunate, and if we can come up with a nice way to avoid or warn about it that would be great, but for what it's worth this is exactly the behavior you want in general from `-r`.

**Comment by shlevy** [comment]
By the way, instead of reinstalling next time you can use one of your profile's old generations to fix.

**Comment by edolstra** [comment]
Well, `-r == --remove-all`, so it can't really violate the principle of least surprise when it in fact removes everything...

BTW, I had to look this flag up, not remembering that it exists. It predates `--set`, which is more appropriate for managing declarative profiles.

**Comment by bryaan** [comment]
@shelvy Yes I realized that after some time, I'm learning. With the nix 2.0 update I think we can make those improvements if they aren't there already.   

@edolstra Technically I agree, maybe it was a poor choice of words, but it stands that it's just too easy to mistype -r and get into trouble.  Now, I didn't realize we had rollback possibilities even on mac, but I am positive I will not be the only user who goes through this same process, and others aren't as patient.  But fingers crossed with 2.0 this is no longer an issue.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
